projects/storefrontlib/src/cms-components/product/product-list/product-facet-navigation/services/facet.service.ts

```diff
@@ -0,0 +1,156 @@
+import { HttpUrlEncodingCodec } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+import { Facet } from '@spartacus/core';
+import { BehaviorSubject, Observable } from 'rxjs';
+import { tap } from 'rxjs/operators';
+import { DialogMode, FacetCollapseState, FacetList } from '../facet.model';
+import { ProductFacetService } from './product-facet.service';
+
+/**
+ * Provides access to the facets as well as UI state for the facets.
+ */
+@Injectable({
+  providedIn: 'root',
+})
+export class FacetService {
+  /** The dialog mode is used to distinquish the experience for mobile vs desktop */
+  protected dialogMode: DialogMode;
+
+  /**
+   * Indicates the last used dialogMode, to differentiate the (default) facet
+   * configuration.
+   */
+  protected lastUsedDialogMode: DialogMode;
+
+  /**
+   * An internal map where we keep the UI state of the facets.
+   */
+  protected facetState = new Map<string, BehaviorSubject<FacetCollapseState>>();
+
+  // will likely be dropped once we move to cxRoute
+  protected queryCodec: HttpUrlEncodingCodec = new HttpUrlEncodingCodec();
+
+  constructor(protected productFacetService: ProductFacetService) {}
+
+  getFacetList(dialogMode: DialogMode): Observable<FacetList> {
+    this.dialogMode = dialogMode;
+    return this.productFacetService.getFacetList().pipe(
+      tap(facetList => {
+        facetList.facets.forEach((facet, position) => {
+          this.configureFacet(facet, position);
+        });
+        this.lastUsedDialogMode = this.dialogMode;
+      })
+    );
+  }
+
+  /**
+   * Returns the UI state for the facet. The state is initialized
+   * using the `initialize` method.
+   */
+  getState(facet: Facet): BehaviorSubject<FacetCollapseState> {
+    this.initialize(facet);
+    return this.facetState.get(facet.name);
```

Comment:
P1: IMO we should not expose full BehaviorSubject, becuase it exposes to consumers the ability to push `.next` values without control. We have 2 options:
- to expose getStateSync (personally I don't like it)
- let consumers do `getState.pipe(take(1)).subscribe(...)`

Reply:
Agreed, good point. I've refactored the code, and also avoid consuming the value directly. This also gets ride of the change detection issue we saw earlier on the host binding for the expanded class.